lib/resolve.go

```diff
@@ -0,0 +1,138 @@
+package lib
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/qri-io/dataset"
+	"github.com/qri-io/qri/base"
+	"github.com/qri-io/qri/base/dsfs"
+	"github.com/qri-io/qri/dsref"
+	"github.com/qri-io/qri/fsi"
+)
+
+// ParseAndResolveRef combines reference parsing and resolution
+func (inst *Instance) ParseAndResolveRef(ctx context.Context, refStr, source string) (dsref.Ref, string, error) {
+	ref, err := dsref.Parse(refStr)
+
+	// bad case references are allowed-but-warned for backwards compatibility
+	if errors.Is(err, dsref.ErrBadCaseName) {
+		log.Error(dsref.ErrBadCaseShouldRename)
+		err = nil
+	} else if err != nil {
+		return ref, "", fmt.Errorf("%q is not a valid dataset reference: %w", refStr, err)
+	}
+
+	resolvedSource, err := inst.ResolveReference(ctx, &ref, source)
+	return ref, resolvedSource, err
+}
+
+// ResolveReference finds the identifier & HEAD path for a dataset reference.
+// the source parameter determines which subsystems of Qri to use when
+func (inst *Instance) ResolveReference(ctx context.Context, ref *dsref.Ref, source string) (string, error) {
```

Comment:
> I guess this is part of my larger dislike of using strings for things that are actually structured data and/or enums.

We can't enumerate all cases at compile time because we want to support exact remote addresses, which in my mind means the set of all possible values to resolve from aren't properly captured by an enum or tuple. The inclination makes a lot of sense to me. Something akin to this:


```go
package lib

type ResolutionMode struct {
  Mode uint8  // specify how to do reference resolution
  Address string //
}

var (
  RMLocal = ResolutionMode{1, ""}
  RMNetwork = ResolutionMode{2, ""}
  RMRegistry = ResolutionMode{3, ""}
)
```

But we have strings coming from packages like `cmd`, which means we end up with something like this:

```
package lib

func ParseResolutionMode(str string) (Mode, error) {
  switch str {
    case "local":
      return RMLocal, nil
    //.. 
    default:
      // TODO - validate address
      return ResolutionMode{4,str}, nil
  }
}
```

So if we have to accept strings, and return Modes, why not just skip straight to strings? In my mind this lets `lib` users skip learning a data structure that can be explained via a comment.

The other issue, because users can specify remote names in addresses, we need configuration to be able to validate an address. If `ParseResolutionMode` from above could be a pure function I'd be more inclined to vote for the mode/enum variation